pvlib/iam.py

```diff
@@ -527,3 +528,214 @@ def sapm(aoi, module, upper=None):
+
+            * 'sky' or 'ground': num = 180
+            * 'horizon': num = 1800
+
+    Returns
+    -------
+    iam : numeric
+        AOI diffuse correction factor for the specified region.
+
+    See Also
+    --------
+    pvlib.iam.marion_diffuse
+
+    References
+    ----------
+    .. [1] B. Marion "Numerical method for angle-of-incidence correction
+       factors for diffuse radiation incident photovoltaic modules",
+       Solar Energy, Volume 147, Pages 344-348. 2017.
+       DOI: 10.1016/j.solener.2017.03.027
+
+    Examples
+    --------
+    >>> marion_integrate(pvlib.iam.ashrae, 20, 'sky')
+    0.9596085829811408
+
+    >>> from functools import partial
+    >>> f = partial(pvlib.iam.physical, n=1.3)
+    >>> marion_integrate(f, [20, 30], 'sky')
+    array([0.96225034, 0.9653219 ])
+    """
+
+    if num is None:
+        if region in ['sky', 'ground']:
+            num = 180
+        elif region == 'horizon':
+            num = 1800
+        else:
+            raise ValueError('Invalid region: {}'.format(region))
+
+    beta = np.radians(surface_tilt)
+    if isinstance(beta, pd.Series):
+        # convert Series to np array for broadcasting later
+        beta = beta.values
+    ai = np.pi/num  # angular increment
+
+    phi_range = np.linspace(0, np.pi, num, endpoint=False)
+    psi_range = np.linspace(0, 2*np.pi, 2*num, endpoint=False)
+
+    # the pseudocode in [1] do these checks at the end, but it's
+    # faster to do this criteria check up front instead of later.
+    if region == 'sky':
+        mask = phi_range + ai <= np.pi/2
+    elif region == 'horizon':
+        lo = 89.5 * np.pi/180
+        hi = np.pi/2
+        mask = (lo <= phi_range) & (phi_range + ai <= hi)
+    elif region == 'ground':
+        mask = (phi_range >= np.pi/2)
+    else:
+        raise ValueError('Invalid region: {}'.format(region))
+    phi_range = phi_range[mask]
+
+    # fast Cartesian product of phi and psi
+    angles = np.array(np.meshgrid(phi_range, psi_range)).T.reshape(-1, 2)
+    # index with single-element lists to maintain 2nd dimension so that
+    # these angle arrays broadcast across the beta array
+    phi_1 = angles[:, [0]]
+    psi_1 = angles[:, [1]]
+    phi_2 = phi_1 + ai
+    # psi_2 = psi_1 + ai  # not needed
+    phi_avg = phi_1 + 0.5*ai
+    psi_avg = psi_1 + 0.5*ai
+    term_1 = np.cos(beta) * np.cos(phi_avg)
+    # we are free to choose any gamma between zero and 2pi, so choose zero
+    # and omit it from the cos(psi_avg) term
+    term_2 = np.sin(beta) * np.sin(phi_avg) * np.cos(psi_avg)
+    aoi = np.arccos(term_1 + term_2)
+    # simplify Eq 8, (psi_2 - psi_1) is always ai
+    dAs = ai * (np.cos(phi_1) - np.cos(phi_2))
+    cosaoi_dAs = np.cos(aoi) * dAs
+    # apply the final AOI check, zeroing out non-passing points
+    mask = aoi < np.pi/2
+    cosaoi_dAs = np.where(mask, cosaoi_dAs, 0)
+    numerator = np.sum(function(np.degrees(aoi)) * cosaoi_dAs, axis=0)
```

Comment:
maybe just for posterity put a comment here explaining why we're purposefully choosing not to use trapezoid rule, like:
```python
# use Riemann sum instead of trapezoid rule to be faithful to Marion's paper,
# number of intervals is large enough that errors are between 1e-4 to 1e-5
```

In all honesty, I lied when I said, "your call." because I think the reason Bill used Reimann instead of trapezoid is because he didn't have access to MATLAB or R? You could probably get away with less intervals and possibly improve the speed by 1 or 2 nanoseconds, AND have your bikeshed painted white! :wink: 😛

Reply:
After thinking about this some more, I think maybe Bill pulled one over on us (or we pulled one over on ourselves) by not actually doing Riemann sums -- he calculates the AOI for the midpoint of each interval (phi_avg and psi_avg), so it's actually a midpoint integration which (if I remember correctly) is even better for smooth functions than the trapezoid rule!  Would you mind taking a second look at the paper when you get a chance and see if you agree?  

Also, I realized my 1e-4/1e-5 differences were from unfair comparisons.  I had just replaced the `np.sum` calls with `np.trapz`, but those numpy arrays are the 2-D integration surface pivoted into a 1-D array, so it doesn't make sense to treat the values as a single sequence of values to integrate.  To do a real comparison we would probably have to redo the method to use the typical meshgrid approach and stick with 2D arrays.  Maybe that's irrelevant now though.

Reply:
fair enough - I agree, midpoint may be better, so no comment required, thanks!